src/ripple/basics/SlabAllocator.h

```diff
@@ -100,91 +209,115 @@ class SlabAllocator
+    std::atomic<SlabBlock*> active_ = nullptr;
 
-                ret = l_;
+    // The size of an item, including the extra bytes requested and
+    // any padding needed for alignment purposes between items:
+    std::size_t const itemSize_;
 
-                if (ret)
-                {
-                    // Use memcpy to avoid unaligned UB
-                    // (will optimize to equivalent code)
-                    std::memcpy(&l_, ret, sizeof(std::uint8_t*));
-                }
-            }
+    // The maximum number of items that this slab can allocate
+    std::size_t const itemCount_;
+
+    // The size of each individual slab:
+    std::size_t const slabSize_;
+
+    [[nodiscard]] std::uint8_t*
+    allocate(SlabBlock* slab) noexcept
+    {
+        std::lock_guard l(slab->m_);
+
+        auto ret = slab->list_;
+
+        if (ret != nullptr)
+        {
+            slab->list_ = ret->next;
+            --slab->avail_;
 
-            return ret;
+            // We want to invoke the destructor for the FreeItem object
+            // before we use its memory.
+            std::destroy_at(ret);
         }
 
-        /** Return an item to this allocator's freelist.
+        return reinterpret_cast<std::uint8_t*>(ret);
+    }
 
-            @param ptr The pointer to the chunk of memory being deallocated.
+    /** Return an item to this allocator's freelist.
+
+        @param slab The slab in which to attempt to return the memory
+        @param ptr The pointer to the chunk of memory being deallocated.
+
+        @return true if the item belongs to the given slab and was freed, false
+                     otherwise.
+    */
+    [[nodiscard]] bool
+    deallocate(SlabBlock* slab, std::uint8_t* ptr) noexcept
+    {
+        assert(ptr != nullptr);
+
+        FreeItem* flp;
 
-            @note This is a dangerous, private interface; the item being
-                  returned should belong to this allocator. Debug builds
-                  will check and assert if this is not the case. Release
-                  builds will not.
-         */
-        void
-        deallocate(std::uint8_t* ptr) noexcept
         {
-            assert(own(ptr));
+            std::lock_guard l(slab->m_);
 
-            std::lock_guard l(m_);
+            if (auto const d = slab->data_; (ptr < d) || (ptr >= d + slabSize_))
+                return false;
 
-            // Use memcpy to avoid unaligned UB
-            // (will optimize to equivalent code)
-            std::memcpy(ptr, &l_, sizeof(std::uint8_t*));
-            l_ = ptr;
-        }
-    };
+            // It looks like this objects belongs to us. Ensure that it
+            // doesn't extend past the end of our buffer:
+            assert(ptr + itemSize_ <= slab->data_ + slabSize_);
 
-private:
-    // A linked list of slabs
-    std::atomic<SlabBlock*> slabs_ = nullptr;
+            slab->list_ = new (std::assume_aligned<alignof(FreeItem)>(ptr))
```

Comment:
isn't it necessary to assert that `ptr` is aligned correctly? I believe `ptr` can only be `SHAMapItem*` as invoked from here: https://github.com/nbougalis/rippled/blob/a4e27672a29d39e26fedd1e074b68b71ef6b2ebd/src/ripple/shamap/SHAMapItem.h#L158

Since `Type` is aligned at 4 or 8 byte boundaries, I feel this should be fine, just wanted to draw your attention for confirmation.

Reply:
Yes, it will be correctly aligned, which is why `std::assume_aligned` is used here. Per the documentation:

>[`std::assume_aligned`] informs the implementation that the object `ptr` points to is aligned to at least `N`. The implementation may use this information to generate more efficient code, but it might only make this assumption if the object is accessed via the return value of `assume_aligned`.

So, we are _communicating_ information to the compiler here that allows it to potentially make better decisions since, by using placement new, it doesn't have control of the alignment of the returned block and so it can't really be sure it's aligned.

That said, it's almost certain that using `std::assume_aligned` doesn't do much and could be removed, but I think it's better to have it, especially considering the pain we go to properly align things.

Reply:
In the Notes section of the cpp-reference: 

> It is up to the program to ensure that the alignment assumption actually holds. A call to assume_aligned does not cause the compiler to verify or enforce this.

`ptr` is obtained as an input to this function. If `ptr` is not aligned to `alignof(FreeItem)`, we could be breaking things. Can we be a 100% sure that `ptr` adheres to this alignment requirements? 

Would an assert like this make the code more robust?
`assert( reinterpret_cast<std::uintptr_t>(ptr) %
                    alignof(FreeItem) ==
                0);`